### ISSUE TITLE
OutcomeTargetRenderer: unwarranted and possibly erroneous inclusion of the context path

### ISSUE THREAD

**Issue by leomekenkamp** [primefaces/primefaces#3506]
## 1) Environment
- PrimeFaces version: 6.2.1
- Does it work on the newest released PrimeFaces version? Version? no
- Does it work on the newest sources in GitHub? (Build by source -> https://github.com/primefaces/primefaces/wiki/Building-From-Source) no
- Application server + version: all / n.a.
- Affected browsers: all / n.a.

## 2) Expected behavior
The url attribute of a MenuItem should not be prepended with facesContext.getExternalContext().getRequestContextPath().

## 3) Actual behavior
If the value of the url attribute starts with '/', the request context path is appended. This leads to problems in the following situations:
1) The request url has been rewritten by a reverse proxy. This will result in a navigation for the browser that is relative not to the externally visible website, but the internal structure behind the proxy. In short: the context path can differ from what the user sees in the url. See https://stackoverflow.com/questions/16675191/get-full-url-and-query-string-in-servlet-for-both-http-and-https-requests/16675399 and scroll down to 'Proxied requests' for more on this.
2) Say our application runs on context root '/foo' and we need a menu item to point to another application on the same server on context root '/bar', we cannot use 'url="/bar/"', because the end result would be http(s)://whateverhost/foo/bar/. Simply adding 'http...' to the url value is not a viable solution, because we do not know the url of our application: our application runs on different hosts.

## Further motivation:
An application can always 'know' on which path relative to its context root a page is on, so it can always use '..' to go one level up. Say we have a page 'subpath/one' in our application. We also have a page 'two' in our application. The context root is unknown. If on page 'subpath/one' we need a link to 'two', we can always specify 'url="../two". There is no reliable way for PrimeFaces or any other framework to get the context root. If an application needs to specify relative to that context root, it needs to do so itself by either using "x-forwarded-for", "x-forwarded-proto", the external context, or whatever works in for their network topology.

## 4) Steps to reproduce
Use any menu item with a url starting with '/'.

## 5) Sample XHTML
<p:menuitem value="abs" url="/" />

## 6) Sample bean
Not needed.

The solution looks simple: remove 'protected String prependContextPathIfNecessary(FacesContext facesContext, String path)' and use the 'path' directly.

**Comment by melloware** [comment]
Related to?  https://github.com/primefaces/primefaces/issues/3154

**Comment by leomekenkamp** [comment]
It is indeed related to #3154. I think the original request cannot be implemented correctly, because in a JSF application it is not possible to know with 100% accuracy what the url in the browser window is. At the same time the JSF/Primefaces code generates a link that should be interpreted by the browser. That means that we can either specify an absolute path (starting with http: or /), or a relative path. That relative path is always interpreted by the browser to be relative to the current path in the url bar.

So we have two disadvantages: 
1. We cannot create a href to another application on the same server without specifying the actual server name.
2. The menu item possibly creates incorrect links when placed behind a reverse proxy.

**Comment by tandraschko** [comment]
how that e.g. h:commandButton handle this?

**Comment by leomekenkamp** [comment]
That is one possibility, but I need the functionality in a menu. And menus need menuitems as children, so I cannot place a commandButton in a menu.

For those who are reading this in hope of a workaround: there is. Simply prepend "/.." to the url attribute. And if your application has a context path that is more than one subpath, simply add a '/..'. 

So if your app is on context path '/bla/foo', add '/../..'.

**Comment by tandraschko** [comment]
i meant how h:commandButton handles the relative url with appending the contextPath e.g.

**Comment by tandraschko** [comment]
Should be fixed, give it a try.

**Comment by firsthour** [comment]
Did this change already sneak into 6.2.6? We got hit by this unexpectedly because we were erroneously using "url" attribute instead of "outcome".  I saw no note of this when I was looking at what changed from 6.2.5 to 6.2.6.

https://github.com/primefaces/primefaces/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aclosed+label%3A6.2.6+

Code in 6.2.5:
`url = prependContextPathIfNecessary(context, href);`

Code in 6.2.6: 
`url = context.getExternalContext().encodeRedirectURL(href, outcomeTarget.getParams());`

I've switched us to using "outcome" but I need to provide an answer to what happened.
